azure-pipelines/cd.yml

```diff
@@ -1,64 +1,174 @@
-    windows:
-      imageName: 'vs2017-win2016'
-
-pool:
-  vmImage: $(imageName)
+parameters:
+- name: buildAgentPoolVar
+  displayName: 'Build agent pool'
+  type: string
+  default: 'BuildAgentPool'
+- name: buildAgentVmImageVar
+  displayName: 'Build agent image'
+  type: string
+  default: 'BuildAgentImage'
+- name: ghRelease
+  displayName: 'Draft github release'
+  type: boolean
+  default: false
+- name: isPreRelease
+  displayName: 'PreRelease / release candidate build'
+  type: boolean
+  default: false
+- name: publishExt
+  displayName: 'Publish to Marketplace'
+  type: boolean
+  default: false
 
 variables:
 - template: common/variables.yml
+- name: vmImage
+  value: $[variables.${{ parameters.buildAgentVmImageVar }}]
+- name: buildPool
+  value: $[variables.${{ parameters.buildAgentPoolVar }}]
+
+stages:
+- stage: test
+  displayName: 'Test'
+  jobs:
+  - job: run_tests
+    displayName: 'Run tests'
+    strategy:
+      matrix:
+        linux:
+          imageName: 'ubuntu-latest'
+        mac:
+          imageName: 'macos-latest'
+        windows:
+          imageName: 'windows-latest'
+    pool:
+      vmImage: $(imageName)
+    steps:
+      - template: common/setup_steps.yml
+      - template: common/compile_steps.yml
+      - template: common/check_steps.yml
+      - template: common/test_steps.yml
+
+- stage: build
+  dependsOn: [test]
+  jobs:
+  - job: CredScan
+    displayName: 'Credential Scan'
+    pool:
+      vmImage: windows-latest
+    steps:
+    - task: CredScan@3
+      inputs:
+        outputFormat: 'pre'
+        scanFolder: '$(Build.SourcesDirectory)'
 
-steps:
+    - task: PostAnalysis@1
+      inputs:
+        AllTools: false
+        APIScan: false
+        BinSkim: false
+        CodesignValidation: false
+        CredScan: true
+        FortifySCA: false
+        FxCop: false
+        ModernCop: false
+        PoliCheck: false
+        RoslynAnalyzers: false
+        SDLNativeRules: false
+        Semmle: false
+        TSLint: false
+        ToolLogsNotFoundAction: 'Standard'
 
-  - template: common/setup_steps.yml
-  - template: common/compile_steps.yml
-  - template: common/check_steps.yml
-  - template: common/test_steps.yml
+  - job: build_and_package
+    displayName: 'Build and Publish Artifacts'
+    pool:
+      name: $(buildPool)
+      vmImage: $(vmImage)
+      demands:
+        - ImageOverride -equals $(vmImage)
+    dependsOn: CredScan
+    steps:
+      - template: common/setup_steps.yml
+      - template: common/compile_steps.yml
+      - template: common/check_steps.yml
+      - template: common/test_steps.yml
 
-  - template: common/set_release_flags_steps.yml
+      # modify application insights key for releases
+      - script: |
+          node scripts/modifyPackageJson.js aiKey $ENV_AIKEY
+        displayName: Modify package.json for releases
+        condition: and(succeeded(), eq(${{ parameters.publishExt }}, True))
+        env:
+          ENV_AIKEY: $(PROD_AIKEY)
```

Comment:
We could add a PROD_AIKEY / targetAiKeyVar parameter -> variable group mapping (like we have for buildAgentPoolVar etc) to explicitly show its a required input and to quickly fulfill it with plugging different var groups.